profiler/src/ProfilerEngine/Datadog.Profiler.Native/ProfileExporter.cpp

```diff
@@ -0,0 +1,719 @@
+        int32_t exportsCount;
+
+        // The goal here is to minimize the amount of time we hold the profileInfo lock.
+        // The lock in ProfileInfoScope guarantees that nobody else is currently holding a reference to the profileInfo.
+        // While inside the lock owned by the profileinfo scope, its profile is moved to the profile local variable
+        // (i.e. the profileinfo will then contains a null profile field when the next sample will be added)
+        // This way, we know that nobody else will ever use that profile again, and we can take our time to manipulate it
+        // outside of the lock.
+        {
+            const auto scope = GetOrCreateInfo(runtimeId);
+
+            // Get everything we need then release the lock
+            profile = std::move(scope.profileInfo.profile);
+            samplesCount = scope.profileInfo.samplesCount;
+
+            // Count is incremented BEFORE creating and sending the .pprof
+            // so that it will be possible to detect "missing" profiles
+            // in the back end
+            exportsCount = ++scope.profileInfo.exportsCount;
+
+            scope.profileInfo.profile = nullptr;
+            scope.profileInfo.samplesCount = 0;
+        }
+
+        const auto& applicationInfo = _applicationStore->GetApplicationInfo(std::string(runtimeId));
+
+        if (profile == nullptr || samplesCount == 0)
+        {
+            Log::Debug("The profiler for application ", applicationInfo.ServiceName, " (runtime id:", runtimeId, ") have empty profile. Nothing will be sent.");
+            continue;
+        }
+
+        if (_exporter == nullptr)
+        {
+            return false;
+        }
+
+        AddProcessSamples(profile.get(), processSamples);
+
+        AddUpscalingRules(profile.get(), upscalingInfos);
+
+        auto additionalTags = libdatadog::Tags{{"env", applicationInfo.Environment},
+                                               {"version", applicationInfo.Version},
+                                               {"service", applicationInfo.ServiceName},
+                                               {"runtime-id", std::string(runtimeId)},
+                                               {"profile_seq", std::to_string(exportsCount - 1)},
+                                               // Optim we can cache the number of cores in a string
+                                               {"number_of_cpu_cores", std::to_string(OsSpecificApi::GetProcessorCount())}};
+
+        if (!applicationInfo.RepositoryUrl.empty())
+        {
+            additionalTags.Add("git.repository_url", applicationInfo.RepositoryUrl);
+        }
+        if (!applicationInfo.CommitSha.empty())
+        {
+            additionalTags.Add("git.commit.sha", applicationInfo.CommitSha);
+        }
+
+        auto filesToSend = std::vector<std::pair<std::string, std::string>>{{MetricsFilename, CreateMetricsFileContent()}};
+        std::string json = GetMetadata();
```

Comment:
The interest of doing it in Send is to reuse the same code to compute the file name instead of overwriting the same files again and again like today. The problem is that we know what to do (i.e. metadata and metrics should be saved as json files) in Export. Maybe it could be done in Send too...